File: source/citnames/source/Output.cc
/*  Copyright (C) 2012-2024 by László Nagy
    This file is part of Bear.

    Bear is a tool to generate compilation database for clang tooling.

    Bear is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    Bear is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "Output.h"
#include "libshell/Command.h"

#include <algorithm>
#include <iomanip>
#include <fstream>
#include <memory>
#include <unordered_set>
#include <utility>

#include <fmt/format.h>
#include <nlohmann/json.hpp>

namespace {

    struct Filter {
        virtual ~Filter() noexcept = default;
        virtual bool apply(const cs::Entry &) = 0;
    };

    struct ContentFilter : public Filter {
        explicit ContentFilter(cs::Content config)
                : config(std::move(config))
        { }

        bool apply(const cs::Entry &entry) override {
            const auto &file = entry.file;
            return exists(file) && to_include(file) && !to_exclude(file);
        }

    private:
        [[nodiscard]] inline bool exists(const fs::path &file) const {
            const auto &to_check = config.include_only_existing_source;
            return (!to_check) || (to_check && does_exist(file));
        }

        [[nodiscard]] inline bool to_include(const fs::path &file) const {
            const auto &include = config.paths_to_include;
            return include.empty() || does_contain(include, file);
        }

        [[nodiscard]] inline bool to_exclude(const fs::path &file) const {
            const auto &exclude = config.paths_to_exclude;
            return !exclude.empty() && does_contain(exclude, file);
        }

        [[nodiscard]] static bool does_exist(const fs::path &path) {
            std::error_code error_code;
            return fs::exists(path, error_code);
        }

        [[nodiscard]] static bool does_contain(const std::list<fs::path> &directories, const fs::path &file) {
            return std::any_of(directories.begin(), directories.end(), [&file](auto directory) {
                // check if the path elements (list of directory names) are the same.
                const auto [end, nothing] = std::mismatch(directory.begin(), directory.end(), file.begin());
                // the file is contained in the directory if all path elements are
                // in the file paths too.
                return (end == directory.end()) || (*end == "");
            });
        }

    private:
        const cs::Content config;
    };

    // Pure version of the boost::hash_combine function.
    static size_t hash_combine(size_t hash, size_t to_combine) {
        return hash ^ (to_combine + 0x9e3779b9 + (hash << 6) + (hash >> 2));
    }

    using DuplicateFilterPtr = std::unique_ptr<struct DuplicateFilter>;

    struct DuplicateFilter : public Filter {
        static DuplicateFilterPtr from_content(const cs::Content&);

        bool apply(const cs::Entry &entry) override {
            const auto h2 = hash(entry);
            auto [_, new_entry] = hashes.emplace(h2);
            return new_entry;
        }

    private:
        virtual size_t hash(const cs::Entry&) const = 0;

        std::unordered_set<size_t> hashes;
    };


    struct FileDuplicateFilter : public DuplicateFilter {
        private:
            size_t hash(const cs::Entry &entry) const override {
                auto string_hasher = std::hash<std::string>{};

                return string_hasher(entry.file);
            }
    };

    struct FileOutputDuplicateFilter : public DuplicateFilter {
        private:
            size_t hash(const cs::Entry &entry) const override {
                auto string_hasher = std::hash<std::string>{};

                auto hash = string_hasher(entry.file);

                if (entry.output) {
                    hash = hash_combine(hash, string_hasher(*entry.output));
                }

                return hash;
            }
    };

    struct StrictDuplicateFilter : public DuplicateFilter {
        private:
            size_t hash(const cs::Entry &entry) const override {
                auto string_hasher = std::hash<std::string>{};

                auto hash = string_hasher(entry.file);

                if (entry.output) {
                    hash = hash_combine(hash, string_hasher(*entry.output));
                }

                for (const auto& arg : entry.arguments) {
                    hash = hash_combine(hash, string_hasher(arg));
                }

                return hash;
            }
    };

    DuplicateFilterPtr DuplicateFilter::from_content(const cs::Content& content) {
        auto fields = content.duplicate_filter_fields;
        if (fields == cs::DUPLICATE_ALL) {
            return std::make_unique<StrictDuplicateFilter>();
        }
        if (fields == cs::DUPLICATE_FILE_OUTPUT) {
            return std::make_unique<FileOutputDuplicateFilter>();
        }
        if (fields == cs::DUPLICATE_FILE) {
            return std::make_unique<FileDuplicateFilter>();
        }

        // If the parameter is invalid use the default filter
        return std::make_unique<FileOutputDuplicateFilter>();
    }

}

namespace cs {

    void validate(const Entry &entry) {
        if (entry.file.empty()) {
            throw std::runtime_error("Field 'file' is empty string.");
        }
        if (entry.directory.empty()) {
            throw std::runtime_error("Field 'directory' is empty string.");
        }
        if (entry.output.has_value() && entry.output.value().empty()) {
            throw std::runtime_error("Field 'output' is empty string.");
        }
        if (entry.arguments.empty()) {
            throw std::runtime_error("Field 'arguments' is empty list.");
        }
    }

    nlohmann::json to_json(const Entry &rhs, const Format &format) {
        nlohmann::json json;

        json["file"] = rhs.file;
        json["directory"] = rhs.directory;
        if (!format.drop_output_field && rhs.output.has_value()) {
            json["output"] = rhs.output.value();
        }
        if (format.command_as_array) {
            json["arguments"] = rhs.arguments;
        } else {
            json["command"] = sh::join(rhs.arguments);
        }

        return json;
    }

    void from_json(const nlohmann::json &j, Entry &entry) {
        j.at("file").get_to(entry.file);
        j.at("directory").get_to(entry.directory);
        if (j.contains("output")) {
            std::string output;
            j.at("output").get_to(output);
            entry.output.emplace(output);
        }
        if (j.contains("arguments")) {
            std::list<std::string> arguments;
            j.at("arguments").get_to(arguments);
            entry.arguments.swap(arguments);
        } else if (j.contains("command")) {
            std::string command;
            j.at("command").get_to(command);

            sh::split(command)
                    .on_success([&entry](auto arguments) {
                        entry.arguments = arguments;
                    })
                    .on_error([](auto error) {
                        throw error;
                    });
        } else {
            throw std::runtime_error("Field 'command' or 'arguments' not found");
        }

        validate(entry);
    }

    bool operator==(const Entry &lhs, const Entry &rhs) {
        return (lhs.file == rhs.file)
               && (lhs.directory == rhs.directory)
               && (lhs.output == rhs.output)
               && (lhs.arguments == rhs.arguments);
    }

    std::ostream &operator<<(std::ostream &os, const Entry &entry) {
        const Format format;
        const nlohmann::json &json = to_json(entry, format);
        os << json;

        return os;
    }

    CompilationDatabase::CompilationDatabase(Format _format, Content _content)
            : format(_format)
            , content(std::move(_content))
    { }


    rust::Result<size_t> CompilationDatabase::to_json(const fs::path &file, const Entries &rhs) const {
        try {
            std::ofstream target(file);
            return to_json(target, rhs)
                    .map_err<std::runtime_error>([&file](auto error) {
                        return std::runtime_error(
                                fmt::format("Failed to write file: {}, cause: {}",
                                            file.string(),
                                            error.what()));
                    });

        } catch (const std::exception &error) {
            return rust::Err(std::runtime_error(
                    fmt::format("Failed to write file: {}, cause: {}",
                                file.string(),
                                error.what())));
        }
    }

    rust::Result<size_t> CompilationDatabase::to_json(std::ostream &ostream, const Entries &entries) const {
        try {
            ContentFilter content_filter(content);
            DuplicateFilterPtr duplicate_filter = DuplicateFilter::from_content(content);

            size_t count = 0;
            nlohmann::json json = nlohmann::json::array();
            for (const auto &entry : entries) {
                if (content_filter.apply(entry) && duplicate_filter->apply(entry)) {
                    auto json_entry = cs::to_json(entry, format);
                    json.emplace_back(std::move(json_entry));
                    ++count;
                }
            }

            ostream << std::setw(2) << json << std::endl;

            return rust::Ok(count);
        } catch (const std::exception &error) {
            return rust::Err(std::runtime_error(error.what()));
        }
    }

    rust::Result<size_t> CompilationDatabase::from_json(const fs::path &file, std::list<cs::Entry> &entries) const {
        try {
            std::ifstream source(file);
            return from_json(source, entries)
                    .map_err<std::runtime_error>([&file](auto error) {
                        return std::runtime_error(
                                fmt::format("Failed to read file: {}, cause: {}",
                                            file.string(),
                                            error.what()));
                    });
        } catch (const std::exception &error) {
            return rust::Err(std::runtime_error(
                    fmt::format("Failed to read file: {}, cause: {}",
                                file.string(),
                                error.what())));
        }
    }

    rust::Result<size_t> CompilationDatabase::from_json(std::istream &istream, std::list<cs::Entry> &entries) const {
        try {
            nlohmann::json in;
            istream >> in;

            for (const auto &e : in) {
                Entry entry;
                cs::from_json(e, entry);
                entries.emplace_back(entry);
            }

            return rust::Ok(in.size());
        } catch (const std::exception &error) {
            return rust::Err(std::runtime_error(error.what()));
        }
    }

    void to_json(nlohmann::json &j, const Entry &entry, const Format &format) {
        j = nlohmann::json{
                {"file",      entry.file},
                {"directory", entry.directory},
        };
        if (!format.drop_output_field && entry.output) {
            j["output"] = entry.output.value();
        }
        if (format.command_as_array) {
            j["arguments"] = entry.arguments;
        } else {
            j["command"] = sh::join(entry.arguments);
        }
    }

    nlohmann::json to_json(const LinkEntry &rhs, const Format &format) {
        nlohmann::json json;
        json["directory"] = rhs.directory;
        json["input_files"] = rhs.input_files;
        if (!format.drop_output_field && rhs.output.has_value()) {
            json["output"] = rhs.output.value();
        }
        if (format.command_as_array) {
            json["arguments"] = rhs.arguments;
        } else {
            json["command"] = sh::join(rhs.arguments);
        }
        return json;
    }

    void to_json(nlohmann::json &j, const LinkEntry &entry, const Format &format) {
        j = nlohmann::json{
                {"directory", entry.directory},
                {"input_files", entry.input_files},
        };
        if (!format.drop_output_field && entry.output.has_value()) {
            j["output"] = entry.output.value();
        }
        if (format.command_as_array) {
            j["arguments"] = entry.arguments;
        } else {
            j["command"] = sh::join(entry.arguments);
        }
    }

    bool operator==(const LinkEntry &lhs, const LinkEntry &rhs) {
        return (lhs.directory == rhs.directory)
               && (lhs.output == rhs.output)
               && (lhs.arguments == rhs.arguments)
               && (lhs.input_files == rhs.input_files);
    }

    std::ostream &operator<<(std::ostream &os, const LinkEntry &entry) {
        const Format format;
        nlohmann::json j;
        to_json(j, entry, format);
        os << j;
        return os;
    }

    rust::Result<size_t> CompilationDatabase::to_link_json(const fs::path &file, const LinkEntries &rhs) const {
        try {
            std::ofstream target(file);
            return to_link_json(target, rhs)
                    .map_err<std::runtime_error>([&file](auto error) {
                        return std::runtime_error(
                                fmt::format("Failed to write file: {}, cause: {}",
                                            file.string(),
                                            error.what()));
                    });

        } catch (const std::exception &error) {
            return rust::Err(std::runtime_error(
                    fmt::format("Failed to write file: {}, cause: {}",
                                file.string(),
                                error.what())));
        }
    }

    rust::Result<size_t> CompilationDatabase::to_link_json(std::ostream &ostream, const LinkEntries &entries) const {
        try {
            size_t count = 0;
            nlohmann::json json = nlohmann::json::array();
            for (const auto &entry : entries) {
                nlohmann::json j;
                cs::to_json(j, entry, format);
                json.emplace_back(std::move(j));
                ++count;
            }

            ostream << std::setw(2) << json << std::endl;

            return rust::Ok(count);
        } catch (const std::exception &error) {
            return rust::Err(std::runtime_error(error.what()));
        }
    }

    rust::Result<size_t> CompilationDatabase::from_link_json(const fs::path &file, LinkEntries &entries) const {
        try {
            std::ifstream source(file);
            return from_link_json(source, entries)
                    .map_err<std::runtime_error>([&file](auto error) {
                        return std::runtime_error(
                                fmt::format("Failed to read file: {}, cause: {}",
                                            file.string(),
                                            error.what()));
                    });
        } catch (const std::exception &error) {
            return rust::Err(std::runtime_error(
                    fmt::format("Failed to read file: {}, cause: {}",
                                file.string(),
                                error.what())));
        }
    }

    rust::Result<size_t> CompilationDatabase::from_link_json(std::istream &istream, LinkEntries &entries) const {
        try {
            nlohmann::json in;
            istream >> in;

            for (const auto &e : in) {
                LinkEntry entry;
                cs::from_json(e, entry);
                entries.emplace_back(std::move(entry));
            }

            return rust::Ok(in.size());
        } catch (const std::exception &error) {
            return rust::Err(std::runtime_error(error.what()));
        }
    }

    void validate(const LinkEntry &entry) {
        if (entry.directory.empty()) {
            throw std::runtime_error("Field 'directory' is empty");
        }
        if (entry.arguments.empty()) {
            throw std::runtime_error("Field 'arguments' is empty");
        }
        if (entry.input_files.empty()) {
            throw std::runtime_error("Field 'input_files' is empty");
        }
        if (entry.output.has_value() && entry.output.value().empty()) {
            throw std::runtime_error("Field 'output' is empty string.");
        }
    }

    void from_json(const nlohmann::json &j, LinkEntry &entry) {
        j.at("directory").get_to(entry.directory);
        j.at("input_files").get_to(entry.input_files);
        if (j.contains("output")) {
            std::string output;
            j.at("output").get_to(output);
            entry.output.emplace(output);
        }
        if (j.contains("arguments")) {
            std::list<std::string> arguments;
            j.at("arguments").get_to(arguments);
            entry.arguments.swap(arguments);
        } else if (j.contains("command")) {
            std::string command;
            j.at("command").get_to(command);

            sh::split(command)
                    .on_success([&entry](auto arguments) {
                        entry.arguments = arguments;
                    })
                    .on_error([](auto error) {
                        throw error;
                    });
        } else {
            throw std::runtime_error("Field 'command' or 'arguments' not found");
        }

        validate(entry);
    }

    void validate(const ArEntry &entry) {
        if (entry.directory.empty()) {
            throw std::runtime_error("Field 'directory' is empty");
        }
        if (entry.arguments.empty()) {
            throw std::runtime_error("Field 'arguments' is empty");
        }
        if (entry.input_files.empty()) {
            throw std::runtime_error("Field 'input_files' is empty");
        }
        if (entry.output.has_value() && entry.output.value().empty()) {
            throw std::runtime_error("Field 'output' is empty string.");
        }
        if (entry.operation.empty()) {
            throw std::runtime_error("Field 'operation' is empty");
        }
    }

    void from_json(const nlohmann::json &j, ArEntry &entry) {
        j.at("directory").get_to(entry.directory);
        j.at("input_files").get_to(entry.input_files);
        j.at("operation").get_to(entry.operation);
        if (j.contains("output")) {
            std::string output;
            j.at("output").get_to(output);
            entry.output.emplace(output);
        }
        if (j.contains("arguments")) {
            std::list<std::string> arguments;
            j.at("arguments").get_to(arguments);
            entry.arguments.swap(arguments);
        } else if (j.contains("command")) {
            std::string command;
            j.at("command").get_to(command);

            sh::split(command)
                    .on_success([&entry](auto arguments) {
                        entry.arguments = arguments;
                    })
                    .on_error([](auto error) {
                        throw error;
                    });
        } else {
            throw std::runtime_error("Field 'command' or 'arguments' not found");
        }

        validate(entry);
    }

    bool operator==(const ArEntry &lhs, const ArEntry &rhs) {
        return (lhs.directory == rhs.directory)
               && (lhs.output == rhs.output)
               && (lhs.arguments == rhs.arguments)
               && (lhs.input_files == rhs.input_files)
               && (lhs.operation == rhs.operation);
    }

    std::ostream &operator<<(std::ostream &os, const ArEntry &entry) {
        const Format format;
        nlohmann::json j;
        to_json(j, entry, format);
        os << j;
        return os;
    }

    void to_json(nlohmann::json &j, const ArEntry &entry, const Format &format) {
        j = nlohmann::json{
                {"directory", entry.directory},
                {"input_files", entry.input_files},
                {"operation", entry.operation}
        };
        if (!format.drop_output_field && entry.output.has_value()) {
            j["output"] = entry.output.value();
        }
        if (format.command_as_array) {
            j["arguments"] = entry.arguments;
        } else {
            j["command"] = sh::join(entry.arguments);
        }
    }

    nlohmann::json to_json(const ArEntry &rhs, const Format &format) {
        nlohmann::json json;
        to_json(json, rhs, format);
        return json;
    }

    rust::Result<size_t> CompilationDatabase::to_ar_json(const fs::path &file, const ArEntries &entries) const {
        try {
            std::ofstream target(file);
            return to_ar_json(target, entries)
                    .map_err<std::runtime_error>([&file](auto error) {
                        return std::runtime_error(
                                fmt::format("Failed to write file: {}, cause: {}",
                                            file.string(),
                                            error.what()));
                    });

        } catch (const std::exception &error) {
            return rust::Err(std::runtime_error(
                    fmt::format("Failed to write file: {}, cause: {}",
                                file.string(),
                                error.what())));
        }
    }

    rust::Result<size_t> CompilationDatabase::to_ar_json(std::ostream &ostream, const ArEntries &entries) const {
        try {
            size_t count = 0;
            nlohmann::json json = nlohmann::json::array();
            for (const auto &entry : entries) {
                nlohmann::json j;
                cs::to_json(j, entry, format);
                json.emplace_back(std::move(j));
                ++count;
            }

            ostream << std::setw(2) << json << std::endl;

            return rust::Ok(count);
        } catch (const std::exception &error) {
            return rust::Err(std::runtime_error(error.what()));
        }
    }

    rust::Result<size_t> CompilationDatabase::from_ar_json(const fs::path &file, ArEntries &entries) const {
        try {
            std::ifstream source(file);
            return from_ar_json(source, entries)
                    .map_err<std::runtime_error>([&file](auto error) {
                        return std::runtime_error(
                                fmt::format("Failed to read file: {}, cause: {}",
                                            file.string(),
                                            error.what()));
                    });
        } catch (const std::exception &error) {
            return rust::Err(std::runtime_error(
                    fmt::format("Failed to read file: {}, cause: {}",
                                file.string(),
                                error.what())));
        }
    }

    rust::Result<size_t> CompilationDatabase::from_ar_json(std::istream &istream, ArEntries &entries) const {
        try {
            nlohmann::json in;
            istream >> in;

            for (const auto &e : in) {
                ArEntry entry;
                cs::from_json(e, entry);
                entries.emplace_back(std::move(entry));
            }

            return rust::Ok(in.size());
        } catch (const std::exception &error) {
            return rust::Err(std::runtime_error(error.what()));
        }
    }
}

File: rust/bear/src/output.rs
/*  Copyright (C) 2012-2024 by László Nagy
    This file is part of Bear.

    Bear is a tool to generate compilation database for clang tooling.

    Bear is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    Bear is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
use std::collections::HashSet;
use std::fs::{File, OpenOptions};
use std::hash::{DefaultHasher, Hash, Hasher};
use std::io::{BufReader, BufWriter};
use std::path::{Path, PathBuf};

use crate::{args, config, filter};
use anyhow::{anyhow, Context, Result};
use json_compilation_db::Entry;
use path_absolutize::Absolutize;
use semantic::{CompilerPass, Meaning};
use serde_json::Error;

/// Responsible for writing the final compilation database file.
///
/// Implements filtering, formatting and atomic file writing.
/// (Atomic file writing implemented by writing to a temporary file and renaming it.)
///
/// Filtering is implemented by the `filter` module, and the formatting is implemented by the
/// `json_compilation_db` module.
pub struct OutputWriter {
    output: PathBuf,
    append: bool,
    filter: config::Filter,
    format: config::Format,
}

impl OutputWriter {
    /// Create a new instance of the output writer.
    pub fn configure(args: &args::BuildSemantic, config: &config::Output) -> Self {
        match config {
            config::Output::Clang { format, filter, .. } => OutputWriter {
                output: PathBuf::from(&args.file_name),
                append: args.append,
                filter: validate_filter(filter),
                format: format.clone(),
            },
            config::Output::Semantic { .. } => {
                todo!("implement this case")
            }
        }
    }

    /// Implements the main logic of the output writer.
    pub fn run(&self, meanings: impl Iterator<Item = semantic::Meaning>) -> anyhow::Result<()> {
        let entries = meanings.flat_map(|value| {
            into_entries(value).unwrap_or_else(|error| {
                log::error!(
                    "Failed to convert semantic meaning to compilation database entries: {}",
                    error
                );
                vec![]
            })
        });
        if self.append && self.output.exists() {
            let from_db = Self::read_from_compilation_db(Path::new(&self.output))?;
            let final_entries = entries.chain(from_db);
            self.write_into_compilation_db(final_entries)
        } else {
            if self.append {
                log::warn!("The output file does not exist, the append option is ignored.");
            }
            self.write_into_compilation_db(entries)
        }
    }

    fn write_into_compilation_db(
        &self,
        entries: impl Iterator<Item = Entry>,
    ) -> anyhow::Result<()> {
        // Filter out the entries as per the configuration.
        let filter: filter::EntryPredicate = TryFrom::try_from(&self.filter)?;
        let filtered_entries = entries.filter(filter);
        // Write the entries to a temporary file.
        self.write_into_temporary_compilation_db(filtered_entries)
            .and_then(|temp| {
                // Rename the temporary file to the final output.
                std::fs::rename(temp.as_path(), self.output.as_path()).with_context(|| {
                    format!(
                        "Failed to rename file from '{:?}' to '{:?}'.",
                        temp.as_path(),
                        self.output.as_path()
                    )
                })
            })
    }

    /// Write the entries to a temporary file and returns the temporary file name.
    fn write_into_temporary_compilation_db(
        &self,
        entries: impl Iterator<Item = Entry>,
    ) -> anyhow::Result<PathBuf> {
        // FIXME: Implement entry formatting.

        // Generate a temporary file name.
        let file_name = self.output.with_extension("tmp");
        // Open the file for writing.
        let file = File::create(&file_name)
            .with_context(|| format!("Failed to create file: {:?}", file_name.as_path()))?;
        // Write the entries to the file.
        json_compilation_db::write(BufWriter::new(file), entries)?;
        // Return the temporary file name.
        Ok(file_name)
    }

    /// Read the compilation database from a file.
    fn read_from_compilation_db(source: &Path) -> anyhow::Result<impl Iterator<Item = Entry>> {
        let file = OpenOptions::new()
            .read(true)
            .open(source)
            .with_context(|| format!("Failed to open file: {:?}", source))?;
        let entries = json_compilation_db::read(BufReader::new(file))
            .flat_map(Self::failed_entry_read_logged);

        Ok(entries)
    }

    fn failed_entry_read_logged(candidate: std::result::Result<Entry, Error>) -> Option<Entry> {
        match candidate {
            Ok(entry) => Some(entry),
            Err(error) => {
                // FIXME: write the file name to the log.
                log::error!("Failed to read entry: {}", error);
                None
            }
        }
    }
}

/// Validate the configuration of the output writer.
///
/// Validation is always successful, but it may modify the configuration values.
fn validate_filter(filter: &config::Filter) -> config::Filter {
    config::Filter {
        source: filter.source.clone(),
        compilers: filter.compilers.clone(),
        duplicates: config::DuplicateFilter {
            by_fields: validate_duplicates_by_fields(
                filter.duplicates.by_fields.as_slice(),
            ),
        },
    }
}

/// Validate the fields of the configuration.
///
/// Removes the duplicates from the list of fields.
fn validate_duplicates_by_fields(fields: &[config::OutputFields]) -> Vec<config::OutputFields> {
    fields
        .into_iter()
        .map(|field| field.clone())
        .collect::<HashSet<_>>()
        .into_iter()
        .collect()
}

impl TryFrom<&config::Filter> for filter::EntryPredicate {
    type Error = anyhow::Error;

    /// Create a filter from the configuration.
    fn try_from(config: &config::Filter) -> std::result::Result<Self, Self::Error> {
        // - Check if the source file exists
        // - Check if the source file is not in the exclude list of the configuration
        // - Check if the source file is in the include list of the configuration
        let source_exist_check = filter::EntryPredicateBuilder::filter_by_source_existence(
            config.source.include_only_existing_files,
        );
        let source_paths_to_exclude = filter::EntryPredicateBuilder::filter_by_source_paths(
            config.source.paths_to_exclude.clone(),
        );
        let source_paths_to_include = filter::EntryPredicateBuilder::filter_by_source_paths(
            config.source.paths_to_include.clone(),
        );
        let source_checks = source_exist_check & !source_paths_to_exclude & source_paths_to_include;
        // - Check if the compiler path is not in the list of the configuration
        // - Check if the compiler arguments are not in the list of the configuration
        let compiler_with_path = filter::EntryPredicateBuilder::filter_by_compiler_paths(
            config.compilers.with_paths.clone(),
        );
        let compiler_with_argument = filter::EntryPredicateBuilder::filter_by_compiler_arguments(
            config.compilers.with_arguments.clone(),
        );
        let compiler_checks = !compiler_with_path & !compiler_with_argument;
        // - Check if the entry is not a duplicate based on the fields of the configuration
        let hash_function = create_hash(config.duplicates.by_fields.clone());
        let duplicates = filter::EntryPredicateBuilder::filter_duplicate_entries(hash_function);

        Ok((source_checks & compiler_checks & duplicates).build())
    }
}

fn create_hash(fields: Vec<config::OutputFields>) -> impl Fn(&Entry) -> u64 + 'static {
    move |entry: &Entry| {
        let mut hasher = DefaultHasher::new();
        for field in &fields {
            match field {
                config::OutputFields::Directory => entry.directory.hash(&mut hasher),
                config::OutputFields::File => entry.file.hash(&mut hasher),
                config::OutputFields::Arguments => entry.arguments.hash(&mut hasher),
                config::OutputFields::Output => entry.output.hash(&mut hasher),
            }
        }
        hasher.finish()
    }
}

pub fn into_entries(value: Meaning) -> Result<Vec<Entry>, anyhow::Error> {
    match value {
        Meaning::Compiler {
            compiler,
            working_dir,
            passes,
        } => {
            let entries = passes
                .iter()
                .flat_map(|pass| -> Result<Entry, anyhow::Error> {
                    match pass {
                        CompilerPass::Preprocess => {
                            Err(anyhow!("preprocess pass should not show up in results"))
                        }
                        CompilerPass::Compile {
                            source,
                            output,
                            flags,
                        } => Ok(Entry {
                            file: into_abspath(source.clone(), working_dir.as_path())?,
                            directory: working_dir.clone(),
                            output: into_abspath_opt(output.clone(), working_dir.as_path())?,
                            arguments: into_arguments(&compiler, source, output, flags)?,
                        }),
                    }
                })
                .collect();

            Ok(entries)
        }
        _ => Ok(vec![]),
    }
}

fn into_arguments(
    compiler: &PathBuf,
    source: &PathBuf,
    output: &Option<PathBuf>,
    flags: &Vec<String>,
) -> Result<Vec<String>, anyhow::Error> {
    let mut arguments: Vec<String> = vec![];
    // Assemble the arguments as it would be for a single source file.
    arguments.push(into_string(&compiler)?);
    for flag in flags {
        arguments.push(flag.clone());
    }
    if let Some(file) = output {
        arguments.push(String::from("-o"));
        arguments.push(into_string(file)?)
    }
    arguments.push(into_string(source)?);
    Ok(arguments)
}

fn into_abspath(path: PathBuf, root: &Path) -> Result<PathBuf, std::io::Error> {
    let candidate = if path.is_absolute() {
        path.absolutize()
    } else {
        path.absolutize_from(root)
    };
    candidate.map(|x| x.to_path_buf())
}

fn into_abspath_opt(path: Option<PathBuf>, root: &Path) -> Result<Option<PathBuf>, std::io::Error> {
    path.map(|v| into_abspath(v, root)).transpose()
}

fn into_string(path: &Path) -> Result<String> {
    path.to_path_buf()
        .into_os_string()
        .into_string()
        .map_err(|_| anyhow!("Path can't be encoded to UTF"))
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::vec_of_strings;

    #[test]
    fn test_non_compilations() -> Result<()> {
        let empty: Vec<Entry> = vec![];

        let result: Vec<Entry> = into_entries(Meaning::Ignored)?;
        assert_eq!(empty, result);

        let input = Meaning::Compiler {
            compiler: PathBuf::from("/usr/bin/cc"),
            working_dir: PathBuf::from("/home/user"),
            passes: vec![],
        };
        let result: Vec<Entry> = into_entries(input)?;
        assert_eq!(empty, result);

        Ok(())
    }

    #[test]
    fn test_single_source_compilation() -> Result<()> {
        let input = Meaning::Compiler {
            compiler: PathBuf::from("clang"),
            working_dir: PathBuf::from("/home/user"),
            passes: vec![CompilerPass::Compile {
                source: PathBuf::from("source.c"),
                output: Some(PathBuf::from("source.o")),
                flags: vec_of_strings!["-Wall"],
            }],
        };

        let expected = vec![Entry {
            directory: PathBuf::from("/home/user"),
            file: PathBuf::from("/home/user/source.c"),
            arguments: vec_of_strings!["clang", "-Wall", "-o", "source.o", "source.c"],
            output: Some(PathBuf::from("/home/user/source.o")),
        }];

        let result: Vec<Entry> = into_entries(input)?;

        assert_eq!(expected, result);

        Ok(())
    }

    #[test]
    fn test_multiple_sources_compilation() -> Result<()> {
        let input = Meaning::Compiler {
            compiler: PathBuf::from("clang"),
            working_dir: PathBuf::from("/home/user"),
            passes: vec![
                CompilerPass::Preprocess,
                CompilerPass::Compile {
                    source: PathBuf::from("/tmp/source1.c"),
                    output: None,
                    flags: vec_of_strings!["-Wall"],
                },
                CompilerPass::Compile {
                    source: PathBuf::from("../source2.c"),
                    output: None,
                    flags: vec_of_strings!["-Wall"],
                },
            ],
        };

        let expected = vec![
            Entry {
                directory: PathBuf::from("/home/user"),
                file: PathBuf::from("/tmp/source1.c"),
                arguments: vec_of_strings!["clang", "-Wall", "/tmp/source1.c"],
                output: None,
            },
            Entry {
                directory: PathBuf::from("/home/user"),
                file: PathBuf::from("/home/source2.c"),
                arguments: vec_of_strings!["clang", "-Wall", "../source2.c"],
                output: None,
            },
        ];

        let result: Vec<Entry> = into_entries(input)?;

        assert_eq!(expected, result);

        Ok(())
    }
}
